include/circt/Dialect/FIRRTL/FIRRTLStructure.td

```diff
@@ -144,6 +144,57 @@ def FExtModuleOp : FIRRTLOp<"extmodule",
   let hasVerifier = 1;
 }
 
+def FMemModuleOp : FIRRTLOp<"memmodule",
+      [IsolatedFromAbove, Symbol, HasParent<"CircuitOp">, OpAsmOpInterface,
+       DeclareOpInterfaceMethods<FModuleLike>,
+       DeclareOpInterfaceMethods<HWModuleLike>]> {
+  let summary = "FIRRTL Generated Module";
+  let description = [{
+    The "firrtl.memmodule" operation represents an external reference to a
+    memory module. See the "firrtl.mem" op for a deeper explantation of the
+    parameters.
+    
+    A "firrtl.mem" operation is typically lowered to this operation when they
+    are not directly lowered to an implementation by the compiler.
+  }];
+  let arguments =
+    (ins UI32Attr:$numReadPorts, UI32Attr:$numWritePorts,
+         UI32Attr:$numReadWritePorts, TypeAttr:$dataType, UI32Attr:$maskBits,
+         UI32Attr:$readLatency, UI32Attr:$writeLatency, UI64Attr:$depth,
+         ArrayAttr:$extraPorts, AnnotationArrayAttr:$annotations);
+  let results = (outs);
+  let regions = (region AnyRegion:$body);
+
+  let skipDefaultBuilders = 1;
+  let builders = [
+    OpBuilder<(ins "StringAttr":$name, "ArrayRef<PortInfo>":$ports,
+                   "uint32_t":$numReadPorts,  "uint32_t":$numWritePorts,
+                   "uint32_t":$numReadWritePorts, 
+                   "FIRRTLType":$dataType, "uint32_t":$maskBits,
+                   "uint32_t":$readLatency, "uint32_t":$writeLatency,
```

Comment:
Do we need constraints similar to `MemOp`
```suggestion
                   Confined<I32Attr, [IntMinValue<0>]>:$readLatency,
                   Confined<I32Attr, [IntMinValue<1>]>:$writeLatency,
```

Reply:
I'm using the unsigned `UI32Attr` variant, so we don't have to worry about adding the constraints. TBH I'm not sure what the best type of attribute to use here is.  I thought that signless integers were the simplest option, but then `getValue()` returns a signed integer and then we started confining the range to `>0`, when really it's up to us to make sure we interpret the integer as unsigned.  So now I'm starting to lean towards making things use unsigned integers instead of signless integers.